### ISSUE TITLE
OSTN p2p spam tests

### ISSUE THREAD

**Issue by mindstyle85** [harmony-one/harmony#2982]
**Describe the bug**

Testing p2p fixes with Edgar:

- 4 rounds so far
- about 4,5 gbps size attack
- shards going down in ~3min but no OOM
- two nil pointer issues found and fixed by edgar

**Comment by mindstyle85** [comment]
just to update:
- we did another attack, first edgars solution didnt work
- he then dived into the attack script code and realised its way more severe than he though
- he is now back to the drawing board and ideas

**Comment by stephen-tse** [comment]
https://github.com/harmony-one/harmony/pull/2984

**Comment by stephen-tse** [comment]
seb's data load: https://gist.githubusercontent.com/SebastianJ/6bd0873eec9afad42d27515ed1bd430b/raw/10a3bd40c791bfbd78d69880ab3cb838c3dd49e5/data.txt

**Comment by stephen-tse** [comment]
on team-devops:

Edgar | HarmonyToday at 3:19 PM
tl;dr there is no need for throttling or atomics , can use more go threads - using order of magnututde more   threads  handles the spam (previously it was atomic that handled it artifically via contention ) , now the problem is liviness of legimtate consensus messages - I will fix that by only accepting consensus messages from messages signed by committee member - this will stop spam from you off the street spammer
I went from 200 go threads to 2500

**Comment by mindstyle85** [comment]
latest updates:
- RJ: https://github.com/harmony-one/harmony/pull/2993
- RJ: the last spamming attack caused the panic, that's why it's stuck
- RJ:  Somehow the nodes are having different heights .I see it's because some of the consensus message got dropped. Some nodes didn't receive some of the committed or prepared msg, so they fall out of sync. The rest of the node carries the consensus forward and then there is not enough nodes that's in sync to reach consensus or view change.

**Comment by SebastianJ** [comment]
We've been running some more tests - network is definitively way more robust now and being able to sustain the attack way longer.

- Shard 2 lost consensus around 8:26 AM - tried a couple of restarts using `./hard-reset.sh -yp os 2 soft_reset`, eventually one of the restarts worked
- Shard 1 lost consensus around 10:04 AM - tried a couple of restarts using `./hard-reset.sh -yp os 1 soft_reset`, still not able to bring it back online

**Comment by SebastianJ** [comment]
Rewrote the spammer to constantly recycle p2p connections - broke shard 0 at 11:59:15 UTC.

New spammer broke the record for egress traffic as well:
![Screenshot 2020-05-13 at 19 06 16](https://user-images.githubusercontent.com/13743/81810671-8be0b000-954d-11ea-9663-fb4419241d69.png)

**Comment by SebastianJ** [comment]
Somehow the case above only manifested for shard 0 - for other shards egress traffic doesn't spin out of control and is throttled. Weird

**Comment by SebastianJ** [comment]
Kept running spam tests on/off between working on other stuff today.

Here's a rough log of events:
Spam started: 05:30 UTC
S0 consensus lost at 06:41:33 May 14 UTC

S1 consensus lost at 06:55:48 May 14 UTC
S1 consensus recovered on its own at 07:05:00 May 14 UTC

S2 consensus lost at 07:07:57 May 14 UTC
S2 consensus recovered on its own at 07:22:21 May 14 UTC

S3 Consensus lost at 08:23:56 May 14 UTC

S1 consensus lost again at 08:53:32 May 14 UTC

S2 consensus lost again at 08:59:55 May 14 UTC

Spammer stopped 9:17 AM UTC

S2 consensus recovered on its own at 09:36:55 May 14 UTC

S2 consensus lost again at 09:41:50 May 14 UTC - ran the spammer for ~3-4 minutes

Spammer stopped 10:00:00 AM UTC

s2 consenus back at 10:18:19 May 14 UTC

--- break ---

s2 consensus spam started 12:54 PM

s2 consenus lost again at 12:55:32 May 14 UTC

s2 consensus recovered on its own at 13:21:49 UTC

Spammer launched targeting s2 ~13:22:00 again

Consensus lost 13:23:32 May 14 UTC

Spammer started 13:55:00 Again

s2 consensus lost at 13:57:52 May 14 UTC

s2 consensus lost again at 14:47:51 May 14 UTC

---

**TLDR**; took down the entire network within a span of ~2-3 hours or so while not really paying attention / optimizing the attack vector.

**Key findings:**
Getting shards to lose consensus during the first spam burst is way harder than ensuing bursts - not really sure why, but current thesis is that Systemd restarts eventually might end up corrupting the databases (either one of the databases, or both, for non-beacon shard nodes) - similar to what happened during the most recent dryrun firedrill.

Shard 0 went down and didn't recover on its own after 1 spam burst.

Shard 1 & 3 required 2 separate spam bursts before not being able to recover on their own.

Shard 2 somehow will always recover when you stop the spam tool - very interesting. Still haven't found a obvious reason in the logs.

![Screenshot 2020-05-14 at 22 18 43](https://user-images.githubusercontent.com/13743/81957262-8e6b0480-9636-11ea-9785-984f2e5b17dd.png)

**Comment by SebastianJ** [comment]
After restarting all shards on OSTN all shards were able to regain consensus 🎉 

![Screenshot 2020-05-14 at 23 27 05](https://user-images.githubusercontent.com/13743/81960172-beb4a200-963a-11ea-8661-1665fce8be5b.png)

**Comment by mindstyle85** [comment]
latest edgar's fix:

- 3-4 attackers with new p2p spam attack
- managed to bring down shard 3 in about 40min
- unsure if shard 2 is down because of an attack or something else happened
